### PR TITLE
Improved .net bindings testing and exception handling

### PR DESCRIPTION
**Pull request by HeDo88TH**
Preparatory for https://github.com/DroneDB/Registry/pull/60

### CODE REVIEW

File: dotnet/DDB.Bindings/DroneDB.cs [hunk 1]
```diff
@@ -34,10 +35,24 @@ static string GetLastError()
 
         public static string Init(string directory)
         {
-            if (_Init(directory, out var outPath) == DDBError.DDBERR_NONE)
-                return Marshal.PtrToStringAnsi(outPath);
+            if (string.IsNullOrWhiteSpace(directory))
```
Comment (pierotofy) [inline dotnet/DDB.Bindings/DroneDB.cs:38]: Should this logic be moved to the C API?

File: dotnet/DDB.Bindings/DroneDB.cs [hunk 2]
```diff
@@ -47,19 +62,39 @@ public static string Init(string directory)
 
         public static List<Entry> Add(string ddbPath, string path, bool recursive = false)
         {
+
```
Comment (pierotofy) [inline dotnet/DDB.Bindings/DroneDB.cs:65]: Same, should this be moved to the lower level of the API?

File: dotnet/DDB.Bindings/DroneDB.cs [hunk 3]
```diff
@@ -47,19 +62,39 @@ public static string Init(string directory)
...
             return Add(ddbPath, new[] { path }, recursive);
+
         }
         public static List<Entry> Add(string ddbPath, string[] paths, bool recursive = false)
         {
-            if (_Add(ddbPath, paths, paths.Length, out var output, recursive) != DDBError.DDBERR_NONE)
-                throw new DDBException(GetLastError());
 
-            var json = Marshal.PtrToStringAnsi(output);
+            if (string.IsNullOrWhiteSpace(ddbPath))
```
Comment (pierotofy) [inline dotnet/DDB.Bindings/DroneDB.cs:75]: Same :)

We do NOT want to add any logic inside the bindings (as much as its feasible anyway) that could then cause different behaviors between other language bindings / command line behavior.

File: dotnet/DDB.Bindings/DroneDB.cs [hunk 4]
```diff
@@ -116,20 +185,42 @@ public static List<Entry> Info(string[] paths, bool recursive = false, int maxRe
...
+                throw new DDBException("DDB path should not be null or empty");
+
+            if (paths == null || paths.Any(string.IsNullOrWhiteSpace))
+                throw new DDBException("One of the provided paths is null or empty");
+
+            try
+            {
+
+                // If the paths are not absolute let's rebase them on ddbPath
+                paths = (from path in paths
```
Comment (pierotofy) [inline dotnet/DDB.Bindings/DroneDB.cs:206]: Mm, I somehow feel this logic should be part of Registry, not the bindings.

**Comment by HeDo88TH** [comment]
I was thinking about the same thing. The bindings should be as "transparent" as possible. I will split the logic between the C implementation and the registry

File: dotnet/DDB.Bindings/DroneDB.cs (hunk 2 above)
Comment (HeDo88TH) [reply dotnet/DDB.Bindings/DroneDB.cs:65]: See https://github.com/DroneDB/DroneDB/pull/115#issuecomment-709561637

File: dotnet/DDB.Bindings/DroneDB.cs (hunk 1 above)
Comment (HeDo88TH) [reply dotnet/DDB.Bindings/DroneDB.cs:38]: See https://github.com/DroneDB/DroneDB/pull/115#issuecomment-709561637

File: dotnet/DDB.Bindings/DroneDB.cs (hunk 3 above)
Comment (HeDo88TH) [reply dotnet/DDB.Bindings/DroneDB.cs:75]: See https://github.com/DroneDB/DroneDB/pull/115#issuecomment-709561637

File: dotnet/DDB.Bindings/DroneDB.cs (hunk 4 above)
Comment (HeDo88TH) [reply dotnet/DDB.Bindings/DroneDB.cs:206]: See https://github.com/DroneDB/DroneDB/pull/115#issuecomment-709561637

**Comment by pierotofy** [comment]
I've moved `cctz` into the ddb core compilation, so that we don't have to statically link cctz.

After that, bindings are still failing with:

```
D       DJI_0027.JPG
  X Info_ImageFile_Details [1s 389ms]
  Error Message:
   DDB.Bindings.DDBException : Error in calling ddb lib. Last error:
 "No matching entries", check inner exception for details
  ----> Newtonsoft.Json.JsonReaderException : Unexpected character e
ncountered while parsing value: {. Path '[0].point_geom', line 1, po
sition 476.
  Stack Trace:
     at DDB.Bindings.DroneDB.Info(String[] paths, Boolean recursive,
 Int32 maxRecursionDepth, Boolean withHash) in /home/runner/work/Dro
neDB/DroneDB/dotnet/DDB.Bindings/DroneDB.cs:line 141
   at DDB.Bindings.DroneDB.Info(String path, Boolean recursive, Int3
2 maxRecursionDepth, Boolean withHash) in /home/runner/work/DroneDB/
DroneDB/dotnet/DDB.Bindings/DroneDB.cs:line 120
   at DDB.Tests.DroneDBTests.Info_ImageFile_Details() in /home/runne
r/work/DroneDB/DroneDB/dotnet/DDB.Tests/DroneDBTests.cs:line 147
--JsonReaderException
   at Newtonsoft.Json.JsonTextReader.ReadStringValue(ReadType readTy
pe)
   at Newtonsoft.Json.JsonTextReader.ReadAsString()
```

But this should be easier to troubleshoot. :+1: 

When the CI fails on Linux, I like to make a temp branch, and add https://github.com/marketplace/actions/debugging-with-tmate into the workflow:

```
- name: Setup tmate session
      uses: mxschmitt/action-tmate@v3
```

Then open a pull request on my own repo (https://github.com/pierotofy/DroneDB/pull/3), which triggers the CI and gives me a shell (this works only on Linux). Then I can troubleshoot the CI environment.

**Comment by HeDo88TH** [comment]
> I've moved `cctz` into the ddb core compilation, so that we don't have to statically link cctz.
> 
> After that, bindings are still failing with:
> 
> ```
> D       DJI_0027.JPG
>   X Info_ImageFile_Details [1s 389ms]
>   Error Message:
>    DDB.Bindings.DDBException : Error in calling ddb lib. Last error:
>  "No matching entries", check inner exception for details
>   ----> Newtonsoft.Json.JsonReaderException : Unexpected character e
> ncountered while parsing value: {. Path '[0].point_geom', line 1, po
> sition 476.
>   Stack Trace:
>      at DDB.Bindings.DroneDB.Info(String[] paths, Boolean recursive,
>  Int32 maxRecursionDepth, Boolean withHash) in /home/runner/work/Dro
> neDB/DroneDB/dotnet/DDB.Bindings/DroneDB.cs:line 141
>    at DDB.Bindings.DroneDB.Info(String path, Boolean recursive, Int3
> 2 maxRecursionDepth, Boolean withHash) in /home/runner/work/DroneDB/
> DroneDB/dotnet/DDB.Bindings/DroneDB.cs:line 120
>    at DDB.Tests.DroneDBTests.Info_ImageFile_Details() in /home/runne
> r/work/DroneDB/DroneDB/dotnet/DDB.Tests/DroneDBTests.cs:line 147
> --JsonReaderException
>    at Newtonsoft.Json.JsonTextReader.ReadStringValue(ReadType readTy
> pe)
>    at Newtonsoft.Json.JsonTextReader.ReadAsString()
> ```
> 
> But this should be easier to troubleshoot. 👍
> 
> When the CI fails on Linux, I like to make a temp branch, and add https://github.com/marketplace/actions/debugging-with-tmate into the workflow:
> 
> ```
> - name: Setup tmate session
>       uses: mxschmitt/action-tmate@v3
> ```
> 
> Then open a pull request on my own repo ([pierotofy#3](https://github.com/pierotofy/DroneDB/pull/3)), which triggers the CI and gives me a shell (this works only on Linux). Then I can troubleshoot the CI environment.

This is gold

**Comment by HeDo88TH** [comment]
https://github.com/DroneDB/DroneDB/pull/115/commits/5d5c2944b17ef8fb88d92d92fdf2e1cbb162a3f6 closes #117

**Comment by HeDo88TH** [comment]
It's missing the parsing of BasicGeometry when reading from database. I will modfy the SELECT query by wrapping point_geom and polygon_geom in AsGeoJSON call. This way we can read the contents and parse the json.
Then I will add a new constructor to BasicPointGeometry and BasicPolygonGeometry that accepts a json as input.
Eventually I can complete the constructor of Entry(Statement& s) by adding the two missing fields: point_geom and polygon_geom.

These changes are needed by Registry because It expects to be able to read the spatial information inside the index. Several tests are failing on https://github.com/DroneDB/Registry/pull/60 because of this.

**Comment by HeDo88TH** [comment]
We hopefully did it @pierotofy

**Review by pierotofy — APPROVED** [review]

**Comment by pierotofy** [comment]
This looks pretty good! No changes required.

Thanks.

**Pull request merged**